tools/sst_rewrite.cc

```diff
@@ -0,0 +1,353 @@
+            }   // if
+
+            // compare two files
+            else
+            {
+                LDbTable file1(options, fname);
+
+                ++cursor;
+                if (NULL!=*cursor)
+                {
+                    fname=*cursor;
+                    LDbTable file2(options, fname);
+
+                    if (file1.GetStatus().ok() && file2.GetStatus().ok())
+                    {
+                        // quick check: same number of keys and bytes of user data?
+                        //     do this before reading entire files
+                        if (file1.GetSstCounter(leveldb::eSstCountKeys)==file2.GetSstCounter(leveldb::eSstCountKeys)
+                            && file1.GetSstCounter(leveldb::eSstCountKeySize)==file2.GetSstCounter(leveldb::eSstCountKeySize)
+                            && file1.GetSstCounter(leveldb::eSstCountValueSize)==file2.GetSstCounter(leveldb::eSstCountValueSize))
+                        {
+                            leveldb::Iterator * it1, *it2;
+                            uint64_t key_count;
+                            bool match;
+
+                            it1=file1.NewIterator();
+                            it2=file2.NewIterator();
+                            match=true;
+
+                            for (it1->SeekToFirst(), it2->SeekToFirst(), key_count=1;
+                                 it1->Valid() && it2->Valid() && match;
+                                 it1->Next(), it2->Next(), ++key_count)
+                            {
+                                match=(0==it1->key().compare(it2->key())) && (0==it1->value().compare(it2->value()));
+
+                                if (!match)
+                                {
+                                    fprintf(stderr, "%s, %s: Content mismatch at key position %d (%d, %d).\n",
+                                            file1.GetFileName(), file2.GetFileName(),
+                                            (int)key_count,
+                                            it1->key().compare(it2->key()), it1->value().compare(it2->value()));
+                                    error_seen=true;
+                                }   // if
+
+                            }   // for
+
+                            if (it1->Valid() != it2->Valid())
+                            {
+                                fprintf(stderr, "%s, %s: Walk of keys terminated early (%d, %d).\n",
+                                        file1.GetFileName(), file2.GetFileName(),
+                                        (int)it1->Valid(), (int)it2->Valid());
+                                error_seen=true;
+                            }
+                        }   // if
+                        else
+                        {
+                            if (file1.GetSstCounter(leveldb::eSstCountKeys)==file2.GetSstCounter(leveldb::eSstCountKeys))
+                                fprintf(stderr, "%s, %s: Number of keys different.\n",
```

Comment:
Several of these error messages can include more information, which would be useful. For example, this message could include the number of keys in each file.